rs/bitcoin/kyt/src/dashboard.rs

```diff
@@ -0,0 +1,107 @@
+use askama::Template;
+use ic_btc_interface::Txid;
+use state::{FetchTxStatus, Timestamp};
+
+#[cfg(test)]
+mod tests;
+
+mod filters {
+    pub fn timestamp_to_datetime<T: std::fmt::Display>(timestamp: T) -> askama::Result<String> {
+        let input = timestamp.to_string();
+        let ts: i128 = input
+            .parse()
+            .map_err(|e| askama::Error::Custom(Box::new(e)))?;
+        let dt_offset = time::OffsetDateTime::from_unix_timestamp_nanos(ts).unwrap();
+        // 2020-12-09T17:25:40+00:00
+        let format =
+            time::format_description::parse("[year]-[month]-[day]T[hour]:[minute]:[second]+00:00")
+                .unwrap();
+        Ok(dt_offset.format(&format).unwrap())
+    }
+}
+
+#[derive(Template)]
+#[template(path = "dashboard.html")]
+pub struct DashboardTemplate {
+    btc_network: BtcNetwork,
+    outcall_capacity: u32,
+    fetch_tx_status: Vec<(Txid, Timestamp, Status)>,
+}
+
+#[derive(Debug, Clone)]
+pub enum Status {
+    PendingOutcall,
+    PendingRetry,
+    Error(String),
+    Fetched(Fetched),
+}
+
+#[derive(Debug, Clone)]
+pub struct Fetched {
+    input_addresses: Vec<Option<bitcoin::Address>>,
+}
+
+impl Status {
+    pub fn to_str(&self) -> &str {
+        match self {
+            Status::PendingOutcall => "Pending outcall",
+            Status::PendingRetry => "Pending retry",
+            Status::Error(_) => "Error",
+            Status::Fetched(_) => "Fetched",
+        }
+    }
+    pub fn is_fetched(&self) -> bool {
+        matches!(self, Status::Fetched(_))
+    }
+    pub fn as_fetched(&self) -> &Fetched {
+        match self {
+            Status::Fetched(fetched) => fetched,
+            _ => panic!("Status is not Fetched"),
+        }
+    }
+    pub fn is_error(&self) -> bool {
+        matches!(self, Status::Error(_))
+    }
+    pub fn as_error(&self) -> &String {
+        match self {
+            Status::Error(err) => err,
+            _ => panic!("Status is not Error"),
+        }
+    }
+}
+
+pub fn dashboard() -> DashboardTemplate {
+    DashboardTemplate {
+        btc_network: state::get_config().btc_network,
+        outcall_capacity: state::OUTCALL_CAPACITY.with(|capacity| *capacity.borrow()),
+        fetch_tx_status: state::FETCH_TX_CACHE.with(|cache| {
+            cache
+                .borrow()
+                .iter()
+                .rev()
+                .map(|(txid, timestamp, status)| {
+                    (
+                        txid,
+                        timestamp,
+                        match status {
+                            FetchTxStatus::PendingOutcall => Status::PendingOutcall,
+                            FetchTxStatus::PendingRetry { .. } => Status::PendingRetry,
+                            FetchTxStatus::Error(err) => Status::Error(format!("{:?}", err)),
+                            FetchTxStatus::Fetched(fetched) => {
+                                // Return an empty list if no input address is available yet.
```

Comment:
I find it a bit confusing that in that case the dashboard displays `Fetched` without any address. Shouldn't the status column somewhat matches the return value of `check_transaction`? What is the return value of `check_transaction` in that case?

Reply:
For all entries in this table, we fetch their transaction data. But we only compute their input addresses when required. The latter is what is directly requested by the `check_transaction` call, and the former is what is figured out by dependency by the canister itself.

It is also why the table is named "Fetch transaction status", not "check_transaction status".

Reply:
Thanks for the explanation! I guess it's still not clear to me in which cases this can occur (not needing to compute input addresses), but this is more a lack of understanding on my side and something we can discuss separately.